### ISSUE TITLE
[Bug]: `<` is incorrectly sanitized within code blocks

### ISSUE THREAD

**Issue by Aloso** [LemmyNet/lemmy#3774]
### Requirements

- [X] Is this a bug report? For questions or discussions use https://lemmy.ml/c/lemmy_support
- [X] Did you check to see if this issue already exists?
- [X] Is this only a single bug? Do not put multiple bugs in one issue.
- [X] Is this a backend issue? Use the [lemmy-ui](https://github.com/LemmyNet/lemmy-ui) repo for UI / frontend issues.

### Summary

When writing a comment like this:

````md
`type` can be used with generic types like `Result<Success, Error>`.

```rust
pub type LibResult = Result<Success, Error>;
```
````

This is "sanitized" to the following:

````md
`type` can be used with generic types like `Result&lt;Success, Error>`.

```rust
pub type LibResult = Result;
```
````

TL;DR: `<` in inline code is replaced with `&lt;`, and generics (which look like an HTML tag) are removed completely.

**NOTE: Sanitization is often discouraged, because it is error prone, as can be observed here. Instead, use a markdown renderer that correctly handles arbitrary, untrusted input!** Most Markdown rendering libraries already disallow HTML, or can be configured to do so. For example, `markdown-it` (used by lemmy-ui) disables HTML by default.

Alternatively, the generated HTML can be sanitized in the client _after converting the markdown to HTML_.

### Steps to Reproduce

1. Write a comment with `<` in inline code or an HTML tag in a code block
2. Submit the comment

### Technical Details

N/A

### Version

0.18.3

### Lemmy Instance URL

https://programming.dev

**Comment by lionirdeadman** [comment]
Likely a bug caused by sanitization here https://github.com/LemmyNet/lemmy/pull/3708

**Comment by Aloso** [comment]
@Nutomic tagging you since you made the PR to sanitize markdown. Note that the [ammonia docs](https://docs.rs/ammonia/3.3.0/ammonia/) say that to render markdown, you should use a markup processor **before** running the sanitizer.

**Comment by RocketDerp** [comment]
See also issue https://github.com/LemmyNet/lemmy/issues/3754 about ampersand in code blocks and titles. I'll try to get your < added to the testing I'm adding.

**Comment by RocketDerp** [comment]
I have test code ready for this situation. It hits on a post body, do I didn't touch the comment testing: https://github.com/RocketDerp/lemmy_server_fixes0/blob/9e02a20db30986c07ceabdb0cf308e12d49da7e3/api_tests/src/post.spec.ts#L536

**Comment by dessalines** [comment]
We've run into so many injection attacks, that back-end sanitization seemed to be the best short term fix to put a stop to them.

The problem in reality isn't markdown-it, but in the hacky way that lemmy-ui adds isomorphic support, which has allowed a ton of xss bugs on initial page load.

The short-term fix is to add *exception rules* to ammonia, to allow certain chars from being sanitized.

The longer term fix, is to either  rework how lemmy-ui does isomorphic support, or start working on lemmy-ui-leptos, which won't have those problems. At that point we can remove ammonia, and back-end sanitization altogether.

**Comment by Aloso** [comment]
@dessalines 

> The short-term fix is to add _exception rules_ to ammonia

I don't see how this could work. To distinguish between a `<` in a code block and a `<` outside a code block, you need a markdown parser. [pulldown-cmark](https://docs.rs/pulldown-cmark/latest/pulldown_cmark/) might be able to do this, although you'd need to manually iterate over the parsed tokens, escape HTML, and convert it back to markdown.

**Comment by Nutomic** [comment]
I think the problem is that ammonia replaces HTML special characters in the first place. This isnt necessary as it already has separate checks to get rid of malicious HTML tags such as `<script>`. I dont see this behaviour documented at all, nor do I see a way to disable it. I opened an upstream issue to see what the maintainers say.

https://github.com/rust-ammonia/ammonia/issues/184

**Comment by Aloso** [comment]
@Nutomic allowing bare `<` characters doesn't completely solve the issue, because the following markdown will still be escaped:

````md
```html
<script>
// some code
</script>
```
````

**Comment by Nutomic** [comment]
I think ammonia generally uses the wrong approach for our use case. Instead I would simply escape HTML special characters:
- `<` becomes `&lt;`
- `&` becomes `&amp;`
- `>` stays as markdown parser handles it

The federation code must leave `&` as is, otherwise it could lead to double escaping so that `&lt;` turns into `&amp;lt` which is wrong.

Alternatively we could get rid of input sanitization entirely and leave that to each frontend. After all, not every frontend uses HTML. For example a mobile app there might be no need to do HTML escaping at all. In this case we would only have to sanitize strings when generating emails and RSS feeds.

**Comment by Aloso** [comment]
@Nutomic simply escaping HTML special characters doesn't fix this bug. `` `<` `` shouldn't be replaced with `` `&lt;` ``, because it is rendered literally as `&lt;`. There's also some subtlety in how titles are rendered.

By the way, here are some examples in the wild:

- https://programming.dev/post/1107671

  There's `&amp;` in the title

- https://programming.dev/post/1297220
  
  Apparently `@rust` in the title is turned into a link, but not rendered as one. Furthermore, the code block contains `-\>`, which might be caused by escaping.

- https://programming.dev/comment/1721320

  This comment contains inline code `&amp;dyn Cat`, which should be `&dyn Cat`

- https://programming.dev/comment/1474895

  Clearly part of the inline code was deleted (this is because of ammonia)

- https://programming.dev/comment/1769978

  This comment shows a ~~terrible~~ beautiful workaround to use `<>` in inline code:

  > **Note:** the `ᐸᐳ` characters are Canadian Aboriginal syllabics because Lemmy devs haven’t fixed broken input sanitization yet.

- https://programming.dev/post/1218909

  In the top comment, `impl<S>` was replaced with `impl<s>` (note the lowercase `s`), and an additional `</s></s>` was added at the end, most likely because of ammonia

- https://programming.dev/post/1347165

  Another example where `&self` is turned into `&amp;self`

**Comment by josephrocca** [comment]
I ran into the `<` problem a couple of days ago, and today I've tried to write this (within triple backticks):

```
<button onclick="foo()">xyz</button>
```
and it gets converted into this:
```
xyz
```
I'm a bit out of the loop here, so I don't know what sort of parsing is causing the whole button text to be removed - imo it shouldn't be removing anything. If a certain element is safe (in terms of XSS), but is not allowed to exist as an actual element on the page for whatever reason, then the sanitizer/parser should have hooks to allow you to escape it, rather than removing it.

**Comment by josephrocca** [comment]
Another weird thing I just noticed:

https://lemmy.world/post/3251517?scrollToComments=true

It looks like what has happened there is the authored post contained `<details>` and some part of Lemmy's parsing/sanitization pipeline has decided to add `</details>` to the end of the post to make sure the post is "valid HTML"? I haven't actually tested this - just going by what I see on that post. HTML parsers do this sort of thing as part of their "never allowed to throw an error" rule.

**Comment by Aloso** [comment]
@dessalines I don't understand how this commit fixes this bug. `<` is still unconditionally replaced with `&lt;`, is it not? So code within backticks will still be incorrectly escaped.

To clarify: This text:

```md
Rust generics use the `<>` characters
```

**must not** be replaced with

```md
Rust generics use the `&lt;>` characters
```

**Comment by Nutomic** [comment]
@Aloso We assume that content is rendered in a browser, so it will automatically display `&lt;` as `<`. If thats not the case eg in apps, its as easy as replacing the escape characters with the original chars.

**Comment by Aloso** [comment]
@Nutomic you are mistaken. This is Markdown, not HTML. In Markdown, `&lt;` is only rendered as `<` when it is not enclosed in a code block or in backticks.

**Comment by Nutomic** [comment]
Then we should not do any escaping inside of code blocks? At the moment its a simple string replace, seems complicated to make it understand markdown.

**Comment by AppleSheeple** [comment]
HTML-sanitizing input markdown has broken discourse in programming communities (and beyond) for weeks now. Not to mention the huge annoyance it causes, and the confusion (or worse, unawareness) new users are faced with due to this brokenness.

I'm struggling to understand how it didn't become clear to devs since day one that doing HTML-sanitization on input markdown is simply wrong. Why is it so hard to just admit that, reverse the change, make recommendations about how to sanitize rendered HTML, implement the recommendations in `lemmy-ui`, and move on?

----

On the plus side, people with a fetish of seeing literal `&amp;`s everywhere are having a blast.

Oh, and some people probably think HKTs are ubiquitously used in Rust's standard library now, since there are plenty of code snippets out there with standalone `Vec`s in them, without the generic part attached. Maybe Lemmy is not the future, because it's from the future!

**Comment by phiresky** [comment]
I agree. It's not possible to use a HTML sanitizer to sanitize markdown because markdown is not HTML, no matter how many replacements / fixes you add on top.

But also, this wasn't added because the core devs looove breaking post content, but because it was the fastest way to fix XSS attacks, which are a larger issue. The proper fix is not as trivial as you seem to imply, but maybe @AppleSheeple you should make a set of PRs against lemmy and lemmy-ui that properly fixes this problem by adding sanitation where it should be?

**Comment by AppleSheeple** [comment]
> I agree. It's not possible to use a HTML sanitizer to sanitize markdown because markdown is not HTML, no matter how many replacements / fixes you add on top.

Maybe I'm reading them wrong, but some of the comments in this issue don't communicate that this is something that is fully grasped.

> But also, this wasn't added because the core devs looove breaking post content, but because it was the fastest way to fix XSS attacks, which are a larger issue. The proper fix is not as trivial as you seem to imply, but maybe @AppleSheeple you should make a set of PRs against lemmy and lemmy-ui that properly fixes this problem by adding sanitation where it should be?

Totally fair. And I'll hold may hands up and admit that maybe I underestimate how webshittery can complicate things. I try to keep myself as far away as possible from that world. So I'm not going to pretend that I know much about it.

Is there a reason why sanitizing in [those functions](https://github.com/LemmyNet/lemmy-ui/blob/d98009c7d50a1349697c972ce8ceb52b1a7244ce/src/shared/markdown.ts#L45-L55) post-render wouldn't simply work?

----

Hopefully I will find some time to contribute something to the Rust part of Lemmy soon(ish).

**Comment by Nutomic** [comment]
@AppleSheeple For one thing, sanitization is also needed for emails and RSS feeds which are rendered by the backend, not lemmy-ui. Also this is meant to help other Lemmy frontends which may not implement any sanitization of their own.

**Comment by dessalines** [comment]
We could sanitize the markdown for emails, but leave it alone for everything else. Primarily this is an issue with lemmy-ui, and how it injects the markdown for isomorphic purposes. That needs to be reworked, then we could remove all this back-end sanitization.
